### PR TITLE
Connect CA Primary Provider refactor

### PR DESCRIPTION
**Pull request by kisunji**
### Description

Best viewed by commit, as the final diff can be confusing.

Key changes:
* Renamed `PrimaryProvider.GenerateRoot` -> `PrimaryProvider.GenerateCAChain`
* Renamed `PrimaryProvider.GenerateIntermediate` -> `PrimaryProvider.GenerateLeafSigningCert`
* Removed `GenerateLeafSigningCert` from `interface PrimaryProvider`
* Folded Intermediate certificate-handling code into Vault's `GenerateCAChain` method to remove leaky abstraction

### Testing & Reproduction steps

No changes to business logic in this PR.

All existing tests are passing.

### Links

Closes https://github.com/hashicorp/consul/issues/12386

### PR Checklist

* [x] updated test coverage
* [x] external facing docs updated
* [x] not a security concern

### CODE REVIEW

File: agent/consul/leader_connect_ca.go [hunk 1]
```diff
@@ -503,39 +510,13 @@ func (c *CAManager) primaryInitialize(provider ca.Provider, conf *structs.CAConf
...
-		c.logger.Info("Correcting stored CARoot values",
-			"previous-signing-key", rootCA.SigningKeyID, "updated-signing-key", expectedSigningKeyID)
-		rootCA.SigningKeyID = expectedSigningKeyID
-		rootUpdateRequired = true
-	}
-
-	// Add the local leaf signing cert to the rootCA struct. This handles both
-	// upgrades of existing state, and new rootCA.
-	if c.getLeafSigningCertFromRoot(rootCA) != interPEM {
-		rootCA.IntermediateCerts = append(rootCA.IntermediateCerts, interPEM)
```
Comment (kisunji) [inline agent/consul/leader_connect_ca.go]: All of this removed logic comes for free in `setLeafSigningCert`.

Comment (kisunji) [reply agent/consul/leader_connect_ca.go]: We lose the logging but I'm not sure how much value it has. Anecdotally I was more confused by it when I first encountered it.

**Review by eikenb — APPROVED** [review]
This looks really good and cleans things up nicely.

Great job on the commits, it really helped reading that way to better understand the changes. It makes me sad that we throw away all that context when we force squash-merge. :disappointed:

**Comment by jkirschner-hashicorp** [comment]
I was also saddened by the forced squash-merges. It makes it difficult to understand past changes if the commits were intentionally curated (to be atomic, meaningful, discrete changes).

**Pull request merged**